# 可选实验 - 正则化代价和梯度

## 目标
在本实验中，你将：
- 将之前的线性和逻辑代价函数扩展为带有正则化项的版本。
- 重新运行之前的过拟合示例，并添加正则化项。


In [1]:
import numpy as np
%matplotlib widget
import matplotlib.pyplot as plt
from plt_overfit import overfit_example, output
from lab_utils_common import sigmoid
np.set_printoptions(precision=8)

# 添加正则化
<img align="Left" src="./images/C1_W3_LinearGradientRegularized.png"  style=" width:400px; padding: 10px; " >
<img align="Center" src="./images/C1_W3_LogisticGradientRegularized.png"  style=" width:400px; padding: 10px; " >

上面的幻灯片显示了线性和逻辑回归的代价和梯度函数。注意：
- 代价
    - 线性和逻辑回归的代价函数差异很大，但向方程中添加正则化是相同的。
- 梯度
    - 线性和逻辑回归的梯度函数非常相似。它们仅在 $f_{wb}$ 的实现上有所不同。

## 带有正则化的代价函数
### 正则化线性回归的代价函数

正则化线性回归的代价函数方程为：
$$J(\mathbf{w},b) = \frac{1}{2m} \sum\limits_{i = 0}^{m-1} (f_{\mathbf{w},b}(\mathbf{x}^{(i)}) - y^{(i)})^2  + \frac{\lambda}{2m}  \sum_{j=0}^{n-1} w_j^2 \tag{1}$$ 
其中:
$$ f_{\mathbf{w},b}(\mathbf{x}^{(i)}) = \mathbf{w} \cdot \mathbf{x}^{(i)} + b  \tag{2} $$ 


将其与不带正则化的代价函数（你在之前的实验中实现过）进行比较，其形式为：

$$J(\mathbf{w},b) = \frac{1}{2m} \sum\limits_{i = 0}^{m-1} (f_{\mathbf{w},b}(\mathbf{x}^{(i)}) - y^{(i)})^2 $$ 

区别在于正则化项 <span style="color:blue">
    $\frac{\lambda}{2m}  \sum_{j=0}^{n-1} w_j^2$ </span>
    
包含这一项可以鼓励梯度下降最小化参数的大小。注意，在这个例子中，参数 $b$ 没有被正则化。这是标准做法。

下面是方程(1)和(2)的实现。请注意，这使用了*本课程的标准模式*，即对所有 `m` 个样本使用 `for 循环`。

In [2]:
def compute_cost_linear_reg(X, y, w, b, lambda_ = 1):
    """
    计算所有样本的代价
    参数:
      X (ndarray (m,n): 数据，m个样本，每个样本有n个特征
      y (ndarray (m,)): 目标值
      w (ndarray (n,)): 模型参数
      b (标量)      : 模型参数
      lambda_ (标量): 控制正则化程度
    返回:
      total_cost (标量): 代价
    """

    m  = X.shape[0]
    n  = len(w)
    cost = 0.
    for i in range(m):
        f_wb_i = np.dot(X[i], w) + b                                   #(n,)(n,)=标量, 参见 np.dot
        cost = cost + (f_wb_i - y[i])**2                               #标量
    cost = cost / (2 * m)                                              #标量
 
    reg_cost = 0
    for j in range(n):
        reg_cost += (w[j]**2)                                          #标量
    reg_cost = (lambda_/(2*m)) * reg_cost                              #标量
    
    total_cost = cost + reg_cost                                       #标量
    return total_cost                                                  #标量


运行下面的单元格查看效果。

In [3]:
np.random.seed(1)
X_tmp = np.random.rand(5,6)
y_tmp = np.array([0,1,0,1,0])
w_tmp = np.random.rand(X_tmp.shape[1]).reshape(-1,)-0.5
b_tmp = 0.5
lambda_tmp = 0.7
cost_tmp = compute_cost_linear_reg(X_tmp, y_tmp, w_tmp, b_tmp, lambda_tmp)

print("Regularized cost:", cost_tmp)

Regularized cost: 0.07917239320214275


**期望输出**:
<table>
  <tr>
    <td> <b>正则化代价: </b> 0.07917239320214275 </td>
  </tr>
</table>

### 正则化逻辑回归的代价函数
对于正则化**逻辑**回归，代价函数的形式为
$$J(\mathbf{w},b) = \frac{1}{m}  \sum_{i=0}^{m-1} \left[ -y^{(i)} \log\left(f_{\mathbf{w},b}\left( \mathbf{x}^{(i)} \right) \right) - \left( 1 - y^{(i)}\right) \log \left( 1 - f_{\mathbf{w},b}\left( \mathbf{x}^{(i)} \right) \right) \right] + \frac{\lambda}{2m}  \sum_{j=0}^{n-1} w_j^2 \tag{3}$$
其中:
$$ f_{\mathbf{w},b}(\mathbf{x}^{(i)}) = sigmoid(\mathbf{w} \cdot \mathbf{x}^{(i)} + b)  \tag{4} $$ 

将其与不带正则化的代价函数（你在之前的实验中实现过）进行比较：

$$ J(\mathbf{w},b) = \frac{1}{m}\sum_{i=0}^{m-1} \left[ (-y^{(i)} \log\left(f_{\mathbf{w},b}\left( \mathbf{x}^{(i)} \right) \right) - \left( 1 - y^{(i)}\right) \log \left( 1 - f_{\mathbf{w},b}\left( \mathbf{x}^{(i)} \right) \right)\right] $$

与上面的线性回归一样，区别在于正则化项 <span style="color:blue">
    $\frac{\lambda}{2m}  \sum_{j=0}^{n-1} w_j^2$ </span>

包含这一项可以鼓励梯度下降最小化参数的大小。注意，在这个例子中，参数 $b$ 没有被正则化。这是标准做法。 

In [4]:
def compute_cost_logistic_reg(X, y, w, b, lambda_ = 1):
    """
    计算所有样本的代价
    参数:
    参数:
      X (ndarray (m,n): 数据，m个样本，每个样本有n个特征
      y (ndarray (m,)): 目标值
      w (ndarray (n,)): 模型参数
      b (标量)      : 模型参数
      lambda_ (标量): 控制正则化程度
    返回:
      total_cost (标量): 代价
    """

    m,n  = X.shape
    cost = 0.
    for i in range(m):
        z_i = np.dot(X[i], w) + b                                      #(n,)(n,)=标量, 参见 np.dot
        f_wb_i = sigmoid(z_i)                                          #标量
        cost +=  -y[i]*np.log(f_wb_i) - (1-y[i])*np.log(1-f_wb_i)      #标量
             
    cost = cost/m                                                      #标量

    reg_cost = 0
    for j in range(n):
        reg_cost += (w[j]**2)                                          #标量
    reg_cost = (lambda_/(2*m)) * reg_cost                              #标量
    
    total_cost = cost + reg_cost                                       #标量
    return total_cost                                                  #标量


运行下面的单元格查看效果。

In [5]:
np.random.seed(1)
X_tmp = np.random.rand(5,6)
y_tmp = np.array([0,1,0,1,0])
w_tmp = np.random.rand(X_tmp.shape[1]).reshape(-1,)-0.5
b_tmp = 0.5
lambda_tmp = 0.7
cost_tmp = compute_cost_logistic_reg(X_tmp, y_tmp, w_tmp, b_tmp, lambda_tmp)

print("Regularized cost:", cost_tmp)

Regularized cost: 0.6850849138741673


**期望输出**:
<table>
  <tr>
    <td> <b>正则化代价: </b> 0.6850849138741673 </td>
  </tr>
</table>

## 带有正则化的梯度下降
运行梯度下降的基本算法不会因正则化而改变，它是：
$$\begin{align*}
&\text{重复直到收敛:} \; \lbrace \\
&  \; \; \;w_j = w_j -  \alpha \frac{\partial J(\mathbf{w},b)}{\partial w_j} \tag{1}  \; & \text{对于 j := 0..n-1} \\ 
&  \; \; \;  \; \;b = b -  \alpha \frac{\partial J(\mathbf{w},b)}{\partial b} \\
&\rbrace
\end{align*}$$
其中每次迭代对所有 $j$ 的 $w_j$ 进行同步更新。

正则化改变的是梯度的计算。

### 带有正则化的梯度计算（线性/逻辑）
线性和逻辑回归的梯度计算几乎相同，仅在 $f_{\mathbf{w}b}$ 的计算上有所不同。
$$\begin{align*}
\frac{\partial J(\mathbf{w},b)}{\partial w_j}  &= \frac{1}{m} \sum\limits_{i = 0}^{m-1} (f_{\mathbf{w},b}(\mathbf{x}^{(i)}) - y^{(i)})x_{j}^{(i)}  +  \frac{\lambda}{m} w_j \tag{2} \\
\frac{\partial J(\mathbf{w},b)}{\partial b}  &= \frac{1}{m} \sum\limits_{i = 0}^{m-1} (f_{\mathbf{w},b}(\mathbf{x}^{(i)}) - y^{(i)}) \tag{3} 
\end{align*}$$

* m 是数据集中的训练样本数量
* $f_{\mathbf{w},b}(x^{(i)})$ 是模型的预测值，而 $y^{(i)}$ 是目标值

      
* 对于 <span style="color:blue"> **线性** </span> 回归模型
    $f_{\mathbf{w},b}(x) = \mathbf{w} \cdot \mathbf{x} + b$
* 对于 <span style="color:blue"> **逻辑** </span> 回归模型
    $z = \mathbf{w} \cdot \mathbf{x} + b$
    $f_{\mathbf{w},b}(x) = g(z)$
    其中 $g(z)$ 是 sigmoid 函数：
    $g(z) = \frac{1}{1+e^{-z}}$
    
添加正则化的项是 <span style="color:blue">$\frac{\lambda}{m} w_j $</span>。

### 正则化线性回归的梯度函数

In [6]:
def compute_gradient_linear_reg(X, y, w, b, lambda_): 
    """
    计算线性回归的梯度
    参数:
      X (ndarray (m,n): 数据，m个样本，每个样本有n个特征
      y (ndarray (m,)): 目标值
      w (ndarray (n,)): 模型参数  
      b (标量)      : 模型参数
      lambda_ (scalar): 控制正则化程度
      
    返回:
      dj_dw (ndarray (n,)): 代价相对于参数w的梯度 
      dj_db (scalar):       代价相对于参数b的梯度 
    """
    m,n = X.shape           #(样本数量, 特征数量)
    dj_dw = np.zeros((n,))
    dj_db = 0.

    for i in range(m):                             
        err = (np.dot(X[i], w) + b) - y[i]                 
        for j in range(n):                         
            dj_dw[j] = dj_dw[j] + err * X[i, j]               
        dj_db = dj_db + err                        
    dj_dw = dj_dw / m                                
    dj_db = dj_db / m   
    
    for j in range(n):
        dj_dw[j] = dj_dw[j] + (lambda_/m) * w[j]

    return dj_db, dj_dw

运行下面的单元格查看效果。

In [ ]:
np.random.seed(1)
X_tmp = np.random.rand(5,3)
y_tmp = np.array([0,1,0,1,0])
w_tmp = np.random.rand(X_tmp.shape[1])
b_tmp = 0.5
lambda_tmp = 0.7
dj_db_tmp, dj_dw_tmp =  compute_gradient_linear_reg(X_tmp, y_tmp, w_tmp, b_tmp, lambda_tmp)

print(f"dj_db: {dj_db_tmp}", )
print(f"Regularized dj_dw:\n {dj_dw_tmp.tolist()}", )

**期望输出**
```
dj_db: 0.6648774569425726
Regularized dj_dw:
 [0.29653214748822276, 0.4911679625918033, 0.21645877535865857]
 ```

### 正则化逻辑回归的梯度函数

In [ ]:
def compute_gradient_logistic_reg(X, y, w, b, lambda_): 
    """
    计算逻辑回归的梯度 
 
    参数:
      X (ndarray (m,n): 数据，m个样本，每个样本有n个特征
      y (ndarray (m,)): 目标值
      w (ndarray (n,)): 模型参数  
      b (标量)      : 模型参数
      lambda_ (scalar): 控制正则化程度
    返回
      dj_dw (ndarray Shape (n,)): 代价相对于参数w的梯度 
      dj_db (scalar)            : 代价相对于参数b的梯度 
    """
    m,n = X.shape
    dj_dw = np.zeros((n,))                            #(n,)
    dj_db = 0.0                                       #标量

    for i in range(m):
        f_wb_i = sigmoid(np.dot(X[i],w) + b)          #(n,)(n,)=scalar
        err_i  = f_wb_i  - y[i]                       #标量
        for j in range(n):
            dj_dw[j] = dj_dw[j] + err_i * X[i,j]      #标量
        dj_db = dj_db + err_i
    dj_dw = dj_dw/m                                   #(n,)
    dj_db = dj_db/m                                   #标量

    for j in range(n):
        dj_dw[j] = dj_dw[j] + (lambda_/m) * w[j]

    return dj_db, dj_dw  


运行下面的单元格查看效果。

In [ ]:
np.random.seed(1)
X_tmp = np.random.rand(5,3)
y_tmp = np.array([0,1,0,1,0])
w_tmp = np.random.rand(X_tmp.shape[1])
b_tmp = 0.5
lambda_tmp = 0.7
dj_db_tmp, dj_dw_tmp =  compute_gradient_logistic_reg(X_tmp, y_tmp, w_tmp, b_tmp, lambda_tmp)

print(f"dj_db: {dj_db_tmp}", )
print(f"Regularized dj_dw:\n {dj_dw_tmp.tolist()}", )

**期望输出**
```
dj_db: 0.341798994972791
Regularized dj_dw:
 [0.17380012933994293, 0.32007507881566943, 0.10776313396851499]
 ```

## 重新运行过拟合示例

In [ ]:
plt.close("all")
display(output)
ofit = overfit_example(True)

在上面的图表中，尝试对之前的示例应用正则化。具体来说：
- 分类（逻辑回归）
    - 将阶数设为6，lambda设为0（无正则化），拟合数据
    - 现在将lambda设为1（增加正则化），拟合数据，注意差异。
- 回归（线性回归）
    - 尝试相同的过程。

## 恭喜！
你已经：
- 获得了线性和逻辑回归中添加正则化的代价和梯度函数的示例
- 对正则化如何减少过拟合有了一些直觉